add perf report blog post (#561)
withastro/astro.build: src/content/blog/web-framework-performance-report.mdx

> FID quantifies a user's experience when trying to interact with unresponsive pages, but it only measures the first interaction. According to Google, INP takes a more well-rounded measurement of a site’s responsiveness by covering a site’s entire spectrum of interactions, from the time a page first begins to load until the user leaves a page. This comprehensive measurement makes INP a more reliable indicator of a site’s overall responsiveness than FID.
>
> The holistic nature of INP makes it more challenging to solve than FID, because your code has to be implemented in a way that protects responsiveness for the user during their entire journey, not just on first load. Since many interactions are done through JavaScript, it means your site has to be loaded carefully for optimized performance.
>
> -- Anne Burnes, [RebelMouse](https://rebelmouse.com/)

While all six of our frameworks struggle with this metric, it's worth noting that the average pass rate across all tracked websites is a suprisingly high 60.9%. While Astro and Wordpress look like the standout successes in the chart above, these sites are performing only slightly above the industry average.

> This will be an interesting metric to watch in 2023, and Google continues to weigh adding FID as an official Core Web Vital.

Reviewer: I know it’s a bit early, so not doing a full review, but flagging this typo:
```suggestion
This will be an interesting metric to watch in 2023, and Google continues to weigh adding INP as an official Core Web Vital.
```

---

> title: "2023 Web Framework Performance Report"

Reviewer 2: Should the URL match the fact that this is intended to be the **2023** Web Framework Performance report? Right now it's just `/blog/web-framework-performance-report`. I'd rename this slug to include `2023`

import CWVFIDChart from "/src/components/charts/CWVFIDChart.astro"
import CWVLCPChart from "/src/components/charts/CWVLCPChart.astro"
import CWVCLSChart from "/src/components/charts/CWVCLSChart.astro"
import CWVINPChart from "/src/components/charts/CWVINPChart.astro"
import LighthouseMedianChart from "/src/components/charts/LighthouseMedianChart.astro"
import LighthousePercentileChart from "/src/components/charts/LighthousePercentileChart.astro"
import CostOfJSChart from "/src/components/charts/CostOfJSChart.astro"

[INTRO]

# The Data

The purpose of this report is to look at real-world data to better understand the relationship between framework choice, performance, and actual user experience on the web. To do this, we looked at three different third-party datasets:

- [The Chrome User Experience Report (CrUX)](https://developer.chrome.com/docs/crux/) provides user experience metrics for how real-world Chrome users experience popular destinations on the web.
- [The HTTP Archive](https://httparchive.org/) which tracks and reports the performance of over 15 million websites over time by regularly collecting Lighthouse performance data.

> - [The Core Web Vitals Technology Report](https://discuss.httparchive.org/t/new-dashboard-the-core-web-vitals-technology-report/2178) which pulls insights from the previous two datasets.

Reviewer 2: Did we actually look at three different data sources, or did we look at the `The Core Web Vitals Technology Report` which combines the data from `CrUX` and `HTTP Archive`?

Author: Only the second two were queried directly, I'll make a note of that

**All data was collected from public, independently-managed datasets.** No performance data was measured directly by the Astro team. All queries used in this report are available at. [Learn more about our methodology.](LINK TO SECTION BELOW)

# The Frameworks

> To create this report, we decided to look at six popular JavaScript-based web frameworks: [Astro](https://astro.build/), [Gatsby](https://gatsbyjs.com/), [Next.js](https://nextjs.org/), [Nuxt](https://nuxtjs.org/), [Remix](https://remix.run/), and [SvelteKit](https://kit.svelte.dev/). We also include data from [Wordpress](https://wordpress.com/) when possible, due to its popularity and gigantic market share on the web.

Reviewer 3: Typo: `Wordpress` -> `WordPress`
```suggestion
To create this report, we decided to look at six popular JavaScript-based web frameworks: [Astro](https://astro.build/), [Gatsby](https://gatsbyjs.com/), [Next.js](https://nextjs.org/), [Nuxt](https://nuxtjs.org/), [Remix](https://remix.run/), and [SvelteKit](https://kit.svelte.dev/). We also include data from [WordPress](https://wordpress.com/) when possible, due to its popularity and gigantic market share on the web.
```

Several exciting new frameworks had to be left out due to not having enough real-world usage in our chosen datasets, but we hope to include more frameworks in the next report.

# Core Web Vitals

[Google’s Core Web Vitals (CWV)](https://web.dev/learn-core-web-vitals/) are a set of three standardized metrics that help you understand how users experience a web page. Each metric measures a different aspect of user experience — load speed, responsiveness, visual stability — and together they quantify the overall performance of a website.

Google’s [Core Web Vitals Assessment](https://pagespeed.web.dev/) is a test that looks at real user measurement data (from the [CrUX](https://developer.chrome.com/docs/crux/) dataset) across all three metrics to determine an overall pass/fail grade for each website. For a website to pass, it must meet the [associated threshold](https://web.dev/defining-core-web-vitals-thresholds/) of **“good”** in all three metrics. If any metric fails the threshold, the website fails the assessment.

> The CWV Assessment is unique in it's use of real-world user data and measurements. This makes it a more accurate reflection of how users actually experience a website, especially over longer sessions. Lighthouse and other lab testing tools are only able to measure first page load.

Reviewer 3: Typo: `it's` -> `its`
```suggestion
The CWV Assessment is unique in its use of real-world user data and measurements. This makes it a more accurate reflection of how users actually experience a website, especially over longer sessions. Lighthouse and other lab testing tools are only able to measure first page load.
```

What is the most likely cause for a website to fail Google's Core Web Vitals Assessment? We can break down the data by individual metric to gain insight into where different frameworks struggle (and succeed) when it comes to web vitals.

<CWVFIDChart />

[First Input Delay (FID)](https://web.dev/fid/) measures the time from when a user first interacts with a page to the time when the browser is able to respond to that interaction. Google's CWV Assessment looks for a FID of 100 milliseconds or less. Anything slower is considered needing improvement and failing the assessment.

Most of our frameworks pass this test handily, with over 90% or more of websites passing the assessment. No framework drops below an 80% pass rate on this test. This means that most of the websites tested are responsive to the first user interaction.

<CWVCLSChart />

[Cumulative Layout Shift (CLS)](https://web.dev/cls/) measures visual stability on the page. To pass this assessment, you should reduce unexpected layout shift to near-zero to give your users a reliable visual experience.

> CLS is an interesting metric for Google to include as one of the three Core Web Vitals because it isn't strictly related to speed or responsiveness. It's inclusion underscores the importance of looking at more than just performance when it comes to measuring the overall quality of user experiences on the web.

Reviewer 3: Typo: `it's` -> `its`
```suggestion
CLS is an interesting metric for Google to include as one of the three Core Web Vitals because it isn't strictly related to speed or responsiveness. Its inclusion underscores the importance of looking at more than just performance when it comes to measuring the overall quality of user experiences on the web.
```

> All frameworks scored 50% or higher in this metric. However, it's the youngest frameworks (Astro, SvelteKit and Remix) that score the highest on this metric. All three scored over 75% on the assessment of this metric across all websites tested.

Reviewer 3: Serial comma
```suggestion
All frameworks scored 50% or higher in this metric. However, it's the youngest frameworks (Astro, SvelteKit, and Remix) that score the highest on this metric. All three scored over 75% on the assessment of this metric across all websites tested.
```

> All frameworks scored 50% or higher in this metric. However, it's the youngest frameworks (Astro, SvelteKit and Remix) that score the highest on this metric. All three scored over 75% on the assessment of this metric across all websites tested.

Reviewer 2: Add footnote about long-tail data?

LCP is the hardest of the three metrics to master. Only 52% of all websites tested pass this metric. Of our six frameworks tested, only Astro and SvelteKit beat this average. The rest come in below the average.

# Coming Soon: Interaction to Next Paint (INP)

**Interaction to Next Paint (INP)** is an experimental web vital that assesses overall website responsiveness, similar to First Input Delay (FID). Where the two metrics differ is that INP observes the latency of all interactions a user has made with the page, not just the first one. It also A low INP means the page was consistently able to respond quickly to all—or the vast majority—of user interactions.

While INP is not an official core web vital today, the Chrome team has signaled [their hope](https://www.youtube.com/watch?v=Mizzbsvv8Os&t=378s) to replace FID with INP as a more holistic, accurate measure of responsiveness.

So, how do our frameworks stack up against this new responsiveness metric?

<CWVINPChart />

> Most noticible in the chart is that a good INP measurement is overall _much_ harder for our frameworks to achieve than First Input Delay (FID). While every framework tested saw an 80%+ pass rate of FID, no framework was able to see that same 80% pass rate on INP. Astro came closest, at 68.8% passing.

Reviewer 3: Typo: `noticible` -> `noticeable`
```suggestion
Most noticeable in the chart is that a good INP measurement is overall _much_ harder for our frameworks to achieve than First Input Delay (FID). While every framework tested saw an 80%+ pass rate of FID, no framework was able to see that same 80% pass rate on INP. Astro came closest, at 68.8% passing.
```

> Most noticible in the chart is that a good INP measurement is overall _much_ harder for our frameworks to achieve than First Input Delay (FID). While every framework tested saw an 80%+ pass rate of FID, no framework was able to see that same 80% pass rate on INP. Astro came closest, at 68.8% passing.

Reviewer 2: A general comment that I'll leave here, I guess.

The phrasing "our frameworks" feels a bit out of place. Would prefer something like "target frameworks" or "the frameworks" rather than "our"

> It's worth noting that the average pass rate across all tracked websites is a suprisingly high 60.9%. While Astro and Wordpress look like the standout successes in the chart above, these sites are actually only performing modestly above the industry average. Why do many of the web frameworks tested struggle with this metric?

Reviewer 3: Typo: `suprisingly` -> `surprisingly`
Typo: `Wordpress` -> `WordPress`
```suggestion
It's worth noting that the average pass rate across all tracked websites is a surprisingly high 60.9%. While Astro and WordPress look like the standout successes in the chart above, these sites are actually only performing modestly above the industry average. Why do many of the web frameworks tested struggle with this metric?
```

> One reason could be that Single Page App (SPA) architecture drives all navigation through JavaScript as a client-side action. This creates an opporunity for input delay that Multi-Page Apps (MPA) without client-side navigation don't have. In an MPA, navigating to a new page triggers a full page load from the server which isn't categorized as input delay. This could help explain why Astro and Wordpress (the two MPAs in the chart) perform significantly better on this metric than the rest of the frameworks tested (all SPAs).

Reviewer 3: Typo: `opporunity` -> `opportunity`
Typo: `Wordpress` -> `WordPress`
```suggestion
One reason could be that Single Page App (SPA) architecture drives all navigation through JavaScript as a client-side action. This creates an opportunity for input delay that Multi-Page Apps (MPA) without client-side navigation don't have. In an MPA, navigating to a new page triggers a full page load from the server which isn't categorized as input delay. This could help explain why Astro and WordPress (the two MPAs in the chart) perform significantly better on this metric than the rest of the frameworks tested (all SPAs).
```

>
> This is particularly difficult on mobile. We took a look at a handful of sites across the industry and within our site network, and found that on mobile INP scores are 35.5% worse than FID on average. When reviewing desktop performance across the same dataset, there was only a 14.1% drop on average.
>
> -- Anne Burnes, [RebelMouse](https://rebelmouse.com/)

This will be an interesting metric to watch in 2023, and Google continues to weigh adding INP as an official Core Web Vital.

# Lighthouse Performance

[Lighthouse](https://developer.chrome.com/docs/lighthouse/overview/) is another tool that we can use to measure the user experience of a website. HTTP Archive runs Lighthouse in simulated mobile loading conditions. This offers a significantly more detailed and consistent analysis of page load performance, down to 100ms fractions of a second. Instead of looking at large “good” vs. “bad” thresholds and buckets, Lighthouse gives you a more detailed performance score measured out of 100.

> Real user data like Core Web Vitals is still the best measurement of the the real user experience, and you can see how the real experience vs. the lab experience differ in some of the charts below. However, there is still interesting insights to be learned from the extra detail that Lighthouse provides. Let's take a look at the data.

Reviewer 3: Double `the`
Typo: `differ` -> `differs`
`is` -> `are`
```suggestion
Real user data like Core Web Vitals is still the best measurement of the real user experience, and you can see how the real experience vs. the lab experience differs in some of the charts below. However, there are still interesting insights to be learned from the extra detail that Lighthouse provides. Let's take a look at the data.
```

<LighthouseMedianChart />

> In the interest of consistency, we have kept the original order from the previous section. However, you will notice that Remix appears much stronger on performance on Lighthouse than it did in the CWV Assessment. One explanation for this may be Remix's use of `startTransition` and `requestIdleCallback` to [defer React hydration on page load.](https://github.com/remix-run/remix/blob/4988875856ba4007d4e1e4219b117ce9c7012489/packages/remix-dev/config/defaults/entry.client.react.tsx) This would theoretically translate to better performance in some lab situations (like Lighthouse) at the expense of increased first-input delay in other, real-world situations.

Reviewer 2: ```suggestion
In the interest of consistency, we have kept the original order from the previous section. However, you will notice that Remix appears much stronger on performance on Lighthouse than it did in the CWV Assessment. One explanation for this may be Remix's use of `startTransition` and `requestIdleCallback` to [defer React hydration on page load.](https://github.com/remix-run/remix/blob/4988875856ba4007d4e1e4219b117ce9c7012489/packages/remix-dev/config/defaults/entry.client.react.tsx) This could theoretically translate to better performance in some lab situations (like Lighthouse) at the expense of increased first-input delay in other, real-world situations.
```

Unfortunately, the median Lighthouse performance score is low across the board. Half of the frameworks tested had a median performance considered "poor" (49 or below) while the other half had a median score that "needs Improvement" (50-89). No frameworks reached a "good" median score of 90+.

Across all tracked websites, the median performance score was a 34/100. To that end, half of our tested frameworks (Astro, SvelteKit, and Remix) did come in above the internet average.

<LighthousePercentileChart />

> By breaking the data down by percentile, we can start to see some slightly more encouraging numbers with Astro and SvelteKit reaching a score of 90+ in the p90 or p95 percentiles. However, the data clearly shows that all websites and frameworks (including Astro) still struggle to achieve good performance in real-world situations.

Reviewer 2: Would add a very explicit callout here that we use the P90 score on the homepage.

Reviewer 2: Also worth calling out _why_ P90 is valuable. Presumably, the reason is because it filters out significant outliers in the dataset.

# **The Cost of JavaScript**

The last thing that we wanted to explore was the relationship between framework choice, performance, and total JavaScript payload size in real-world usage. Do the fastest frameworks tend to be the ones that also send the least amount of JavaScript to the client?

> <CostOfJSChart />

Reviewer 2: Persistent labels would really help on this chart, rather than just ones that appear on hover

> The trend in the data is clear: sites that ship less JavaScript tend to perform better. However, there are too many factor at play for us to confidently tie this trend back to the choice of web framework itself. It could be the case that some frameworks encourage/discourage JavaScript differently than others, but more research is needed before we draw any conclusions.

Reviewer 2: ```suggestion
The trend in the data is clear: sites that ship less JavaScript tend to perform better. However, there are too many factors at play for us to confidently tie this trend back to the choice of web framework itself. It could be the case that some frameworks encourage/discourage JavaScript differently than others, but more research is needed before we draw any conclusions.
```

# Methodology & Limitations

This report was compiled from several publicly available datasets. You can learn more about these datasets and their methodology here: [HTTP Archive methodology](https://almanac.httparchive.org/en/2021/methodology#dataset), [CrUX methodology](https://developer.chrome.com/docs/crux/methodology/), and [CWV Technology Report methodology](https://discuss.httparchive.org/t/new-dashboard-the-core-web-vitals-technology-report/2178).

> Due to capacity limitations, our analysis only looks at the homepages of every tracked website. One benefit of this limitation is that there is less variance in the purpose and use-case of each analyzed website. However, on drawback is that this also means interior pages (like `/about` and `/admin/...` pages) and the technologies that they use are not analyzed and therefore excluded from our analysis.

Reviewer 3: I think it should be "use case" (without a hyphen)? 
Typo: `on` -> `one`
```suggestion
Due to capacity limitations, our analysis only looks at the homepages of every tracked website. One benefit of this limitation is that there is less variance in the purpose and use case of each analyzed website. However, one drawback is that this also means interior pages (like `/about` and `/admin/...` pages) and the technologies that they use are not analyzed and therefore excluded from our analysis.
```

> Another limitation that is unexplored in this report is the impact of a framework's age on measured web performance. The older frameworks measured here (Gatsby, Next.js, Nuxt) have a longer tail of legacy websites running older versions of their framework which are included in the dataset. This creates a situation where only the newer frameworks (Astro, Remix, SvelteKit) can be assumed to be running more modern versions of their software from the last 1-2 years, which could impact the data. This is a limitation of the data that we have available, but it is something that we hope to explore in future reports.

Reviewer 2: Is it worth moving methodology up to the top? These seem like good things to keep in mind _while_ reading the data, not _after_ reading the data.

Author: I link to it in the "data" section, but it's a pretty big section to include at the top. The fear is the opposite and that it's intimidating, getting in the way of the reader trying to see the data. Let me see about making the link more prevalent.
